common/overlay/overlay.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package overlay
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"syscall"
+
+	"github.com/coreos/rkt/pkg/label"
+	"github.com/hashicorp/errwrap"
+)
+
+// Mount mounts a pod's upper and lower directories to the destination directory
+func Mount(lowerDir, treeStoreID, cdir, dest, appName, lbl string,
+	gid int, fm os.FileMode) (string, error) {
+	fi, err := os.Stat(lowerDir)
+	if err != nil {
+		return "", err
+	}
+	imgMode := fi.Mode()
+
+	destRootfs := path.Join(dest, "rootfs")
+	if err := os.MkdirAll(destRootfs, imgMode); err != nil {
+		return "", err
+	}
+
+	overlayDir := path.Join(cdir, "overlay")
+	if err := os.MkdirAll(overlayDir, fm); err != nil {
+		return "", err
+	}
+
+	// Since the parent directory (rkt/pods/$STATE/$POD_UUID) has the 'S_ISGID' bit, here
+	// we need to explicitly turn the bit off when creating this overlay
+	// directory so that it won't inherit the bit. Otherwise the files
+	// created by users within the pod will inherit the 'S_ISGID' bit
+	// as well.
+	if err := os.Chmod(overlayDir, fm); err != nil {
+		return "", err
+	}
+
+	imgDir := path.Join(overlayDir, treeStoreID)
+	if err := os.MkdirAll(imgDir, fm); err != nil {
+		return "", err
+	}
+
+	// Also make 'rkt/pods/$STATE/$POD_UUID/overlay/$IMAGE_ID' to be readable by 'rkt' group
+	// As 'rkt' status will read the 'rkt/pods/$STATE/$POD_UUID/overlay/$IMAGE_ID/upper/rkt/status/$APP'
+	// to get exgid
+	if err := os.Chown(imgDir, -1, gid); err != nil {
```

Comment:
The Stat/Chown/Chmod/MkdirAll/SetFileLabel are not useful when `overlay.Mount()` is called from the rkt-export command because the the overlay dirs are supposed to already be prepared. Could this be factored back to `overlayRender()`?